electronics_lib/PowerConditioning.py

```diff
@@ -263,3 +263,107 @@ def connected_from(self, gnd: Optional[Port[VoltageLink]] = None, pwr_hi: Option
+  on by an external signal, and provides the button output as a signal.
+
+  Attributes:
+    pull_resistance: The resistance of the pull-up resistor.
+    amp_resistance: The resistance of the amplifier resistor.
+    diode_drop: The voltage drop across the diode.
+
+  Ports:
+    pwr_in: A voltage sink port for the input power.
+    pwr_out: A voltage source port for the output power.
+    gnd: ground.
+    btn_out: A digital single source port for the button output signal.
+    btn_in: A digital single source port for the button input signal. Should be connected to a button. Do not connect IO.
+    control: A digital sink port for external control to latch the power on.
+  """
+  @init_in_parent
+  def __init__(self, pull_resistance: RangeLike=10*kOhm(tol=0.05), amp_resistance: RangeLike=10*kOhm(tol=0.05), diode_drop:RangeLike=(0, 0.4)*Volt):
+    super().__init__()
+    self.pwr_in = self.Port(VoltageSink.empty(), [Input])
+    self.pwr_out = self.Port(VoltageSource.empty(), [Output])
+    self.gnd = self.Port(Ground.empty(), [Common])
+
+    self.btn_out = self.Port(DigitalSingleSource.empty())
+    self.btn_in = self.Port(DigitalSingleSource.empty())
+    self.control = self.Port(DigitalSink.empty())  # digital level control - gnd-referenced NFET gate
+
+    self.pull_resistance = self.ArgParameter(pull_resistance)
+    self.amp_resistance = self.ArgParameter(amp_resistance)
+    self.diode_drop = self.ArgParameter(diode_drop)
+
+  def contents(self):
+    super().contents()
+
+    control_voltage = self.control.link().voltage.hull(RangeExpr.ZERO,)
+    pwr_voltage = self.pwr_out.link().voltage.hull(RangeExpr.ZERO,)
+    pwr_current = self.pwr_out.link().current_drawn.hull(RangeExpr.ZERO,)
+
+    self.pull_res = self.Block(Resistor(
+      resistance=self.pull_resistance
+    ))
+    self.pwr_fet = self.Block(Fet.PFet(
+      drain_voltage=pwr_voltage,
+      drain_current=pwr_current,
+      gate_voltage=(control_voltage.upper(), control_voltage.upper()),  # TODO this ignores the diode drop
+    ))
+
+    self.amp_res = self.Block(Resistor(
+      resistance=self.amp_resistance
+    ))
+    self.amp_fet = self.Block(Fet.NFet(
+      drain_voltage=control_voltage,
+      drain_current=RangeExpr.ZERO,  # effectively no current
+      gate_voltage=(self.control.link().output_thresholds.upper(), self.control.link().voltage.upper())
+    ))
+
+    self.ctl_diode = self.Block(Diode(
+      reverse_voltage=control_voltage,
+      current=RangeExpr.ZERO,  # effectively no current
+      voltage_drop=self.diode_drop,
+      reverse_recovery_time=RangeExpr.ALL
+    ))
+
+    self.btn_diode = self.Block(Diode(
+      reverse_voltage=control_voltage,
+      current=RangeExpr.ZERO,  # effectively no current
+      voltage_drop=self.diode_drop,
+      reverse_recovery_time=RangeExpr.ALL
+    ))
+
+    self.import_kicad(self.file_path("resources", f"{self.__class__.__name__}.kicad_sch"),
```

Comment:
I think the schematic still has the switch in it? Does that need to get removed?

Reply:
I think it should be removed already

![image](https://github.com/BerkeleyHCI/PolymorphicBlocks/assets/49264928/6fb2f329-01fe-4278-98be-418714f12116)